### PR TITLE
Harmony 933 concatenate parameter

### PR DESCRIPTION
**Pull request by indiejames**
Adds the 'concatenate' parameter along with OpenAPI validation, etc., but only enables it for tests for now since the actual code to run the CONCISE service will not be added until HARMONY-959-662.

### CODE REVIEW

File: app/middleware/earthdata-login-token-authorizer.ts [hunk 1]
```diff
@@ -92,6 +92,12 @@ export default function buildEdlAuthorizer(paths: Array<string | RegExp> = []):
           next(e);
         }
       }
+    } else if (req.headers.cookie) {
```
[the remark this answers is not in the document]
Comment (indiejames) [reply app/middleware/earthdata-login-token-authorizer.ts]: Fixed with help from @bilts . Removed this code since it is unnecessary.

**Review by chris-durbin — APPROVED** [review]

File: app/models/services/index.ts [hunk 2]
```diff
@@ -303,7 +321,35 @@ class UnsupportedOperation extends Error {
...
+ * @returns Any service configurations that support concatenation
+ */
+function filterConcatenationMatches(
+  operation: DataOperation,
+  context: RequestContext,
+  configs: ServiceConfig<unknown>[],
+  requestedOperations: string[],
+): ServiceConfig<unknown>[] {
+  let matches = configs;
+  if (requiresConcatenation(operation)) {
```
[the remark this answers is not in the document]
Comment (bilts) [reply app/models/services/index.ts:350]: I disagree with Chris :-/ at least as written.  If the _only_ capability is concatenation, maybe we skip it, but PO.DAAC will usually be chaining CONCISE with other services, so in those cases, we'll want to use the flag to run / skip CONCISE based on whether the param was set.

File: app/models/services/index.ts [hunk 3]
```diff
@@ -627,6 +677,10 @@ export function chooseServiceConfig(
   configs: ServiceConfig<unknown>[] = serviceConfigs,
 ): ServiceConfig<unknown> {
   let serviceConfig = filterServiceConfigs(operation, context, configs, allFilterFns);
+  // if we are asked to concatenate, but no matching concat service is available then throw an error
+  if (serviceConfig.name === 'noOpService' && operation.shouldConcatenate) {
```
[the remark this answers is not in the document]
Comment (bilts) [reply app/models/services/index.ts:691]: I'll just go ahead and agree that unprocessed links are weird.

File: app/routers/router.ts [hunk 4]
```diff
@@ -92,6 +92,7 @@ function service(fn: RequestHandler): RequestHandler {
       if (!req.collections || req.collections.length === 0) {
         throw new NotFoundError('Services can only be invoked when a valid collection is supplied in the URL path before the service name.');
       }
+      console.log('ok');
```
Comment (bilts) [inline app/routers/router.ts:95]: Remove (this should be a linter issue)

File: test/concatenation.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+        expect(res.statusCode).to.equal(400);
+        expect(res.body).to.eql({
+          code: 'harmony.RequestValidationError',
+          description: 'Error: query parameter "concatenate" must be \'false\' or \'true\'',
+        });
+      });
+    });
+
+    describe('is requested for a collection that has no service that provides concatenation', function () {
```
[the remark this answers is not in the document]
Comment (bilts) [reply test/concatenation.ts:99]: See test/models/services.ts. - we have ways to provide an explicit alternative service config set to whatever we want.

**Review by bilts — CHANGES_REQUESTED** [review]
I think the logic for `supportsOperationsOtherThanConcatenation` needs to change or the use of that method needs to be removed to allow for cases where PO.DAAC wants to chain services into CONCISE and have concatenation be an optional part of that.

Note: the ticket number is 993 not 933.  Please make sure commit messages have that number moving forward to help with release notes.

File: app/models/services/index.ts [hunk 6]
```diff
@@ -120,6 +120,25 @@ export function buildService(
...
+}
+
+/**
+ * Returns any services that don't support concatenation from the list of configs
+ * @param configs - The potential matching service configurations
+ * @returns any configurations that don't support concatenation
+ */
+function supportsOperationsOtherThanConcatenation(
+  configs: ServiceConfig<unknown>[]): ServiceConfig<unknown>[] {
+  return configs.filter((config) => !getIn(config, 'capabilities.concatenation', false));
```
Comment (chris-durbin) [inline app/models/services/index.ts:139]: Oh reading this logic I see the issue. The intent is that we should not send a request to a service chain that requires concatenation. That should not be based on the `capabilities`. If we want to revert this change and write a separate ticket I'd be good with that.

File: app/routers/router.ts (hunk 4 above)
Comment (indiejames) [reply app/routers/router.ts:95]: Done. Should not have been committed.

File: test/concatenation.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (chris-durbin) [reply test/concatenation.ts:99]: We should defer this test to the new ticket which will make sure that services that would skip concatenation are not used.

**Review by bilts — APPROVED** [review]

**Pull request merged**